### ISSUE TITLE
For bounded ranges, require that end is reachable (finitely?) from begin (?)

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/stl2#63]
Needs more thought. Discussion [here](https://github.com/ericniebler/range-v3/pull/183#issuecomment-120771153)

**Comment by CaseyCarter** [comment]
A bounded range has traditionally been defined as delimited by a begin iterator `i` and a past-the-end iterator `j`, denoted `[i, j)`. For a finite range with cardinality N, incrementing the begin iterator N times results in the end iterator: `i + N == j`. The logical extension of this to infinite ranges would be to require that `j` is the limit of the sequence of successive applications of increment to `i`, which would be hard to specify/verify ;)

On a pragmatic note, there are millions of lines of code out there that believe that bounded ranges are finite, and that `std::distance(i, j)` determines their cardinality. If we decide to break that assumption the fallout will be massive.

Now that we have the notion of ranges denoted by differently typed iterators and sentinels, relaxing the requirement that bounded ranges are finite doesn't expand the class of ranges that we can describe. I can always pair an iterator `i` with `unreachable` to describe the infinite range of elements beginning at `i` so I don't need unbounded bounded ranges.

**Comment by ericniebler** [comment]
> On a pragmatic note, there are millions of lines of code out there that believe that bounded ranges are finite

Are you sure? For the history of the STL, there has been no other way to represent ranges besides a pair of iterators of the same type. You don't think people have (ab)used this to represent infinite ranges? Even pairs of `std::istream_iterator` represents a (potentially) infinite range.

Taking an iterator `I` and `unreachable` and bundling then into a pair of `common_iterator<I, unreachable>` results in an unbounded, bounded range. Such a common iterator cannot be stronger than `Forward`. Certainly Bounded *Bidirectional* ranges can guarantee the property you describe, but it seems to be that Bounded *Forward* ranges cannot.

**Comment by CaseyCarter** [comment]
A bounded Bidirectional range could be infinite if the begin/end iterators never meet in the middle - the orbits of the begin iterator under ++ and the end iterator  under -- would be independent. I suppose there's an equivalent formulation for random access iterators as well, if begin + N is less than end - N for all N.

**Comment by CaseyCarter** [comment]
[Infinite bounded random access range](https://gist.github.com/CaseyCarter/a4aadef0413e826cc7be):
```cpp
#include <range/v3/view_facade.hpp>
#include <iostream>

using namespace ranges;

template <class T>
class weirdrange :
  public view_facade<weirdrange<T>, cardinality::infinite> {
  friend class ranges::range_access;

  CONCEPT_ASSERT(UnsignedIntegral<T>());
  T first_, last_;

  class cursor {
    friend class ranges::range_access;
    using D = meta::_t<std::make_signed<T>>;

    T pos_;

    T current() const { return pos_; }
    void next() { pos_ += 2; }
    bool equal(cursor other) const { return pos_ == other.pos_; }
    void prev() { pos_ -= 2; }
    void advance(D n) { pos_ += 2 * static_cast<T>(n); }
    D distance_to(cursor other) const { return (other.pos_ - pos_) / 2; }
  public:
    cursor() = default;
    cursor(T pos) :
      pos_{pos} {}
  };

  cursor begin_cursor() const { return first_; }
  cursor end_cursor() const { return last_; }

public:
  weirdrange() = default;
  weirdrange(T first, T last) :
    first_{first}, last_{last} {}
};

int main() {
  auto r = weirdrange<unsigned char>{0,1};
  CONCEPT_ASSERT(BidirectionalView<decltype(r)>());
  CONCEPT_ASSERT(RandomAccessView<decltype(r)>());

  std::cout << std::hex << std::showbase;
  auto i = begin(r);
  auto e = end(r);
  while (i != e) {
    std::cout << (unsigned)*i++ << ' ' << (unsigned)*--e << '\n';
  }
}
```

**Comment by ericniebler** [comment]
> `weirdrange<uint8_t>{0,1}` is infinite since the orbits of `begin` and `end` are independent

I wasn't doubting that it's possible to construct such weird ranges. I was wondering whether such ranges are valid.

**Comment by ericniebler** [comment]
Since it's not clear to me that `Bounded` implies finite and acyclic, I'm inclined to just close this bug as invalid. Any last arguments?

**Comment by CaseyCarter** [comment]
> Any last arguments?

I think we've discussed everything pertinent. If we're going to allow infinite `BoundedRange`, we should probably do something with "Range `[i,j)` is valid if and only if `j` is reachable from `i`. The result of the application of functions in the library to invalid ranges is undefined." in [iterator.requirements.general]/9.

**Comment by ericniebler** [comment]
Yeah. The reachability thing is awkward. Really, infinite ranges need major
work and research. The core concepts need to change to avoid overflow on
difference type.

\e

**Comment by CaseyCarter** [comment]
I'm pushing this out to IS: it would be nice to have infinite ranges, but I don't think we'll get there for the TS.
